iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp

```diff
@@ -72,11 +88,61 @@ class ServiceRegistry
     const ServiceDescriptionVector_t getServices() const noexcept;
 
   private:
-    /// @todo #859 replace std::multimap with prefix tree
-    ::std::multimap<capro::IdString_t, uint64_t> m_serviceMap;
-    ::std::multimap<capro::IdString_t, uint64_t> m_instanceMap;
-    ::std::multimap<capro::IdString_t, uint64_t> m_eventMap;
-    ServiceDescriptionVector_t m_serviceDescriptionVector;
+    using Entry_t = cxx::optional<ServiceDescriptionEntry>;
+    using ServiceDescriptionContainer_t = cxx::vector<Entry_t, MAX_SERVICE_DESCRIPTIONS>;
+
+    static constexpr uint32_t NO_INDEX = MAX_SERVICE_DESCRIPTIONS;
+
+    // tag type for internal overloads
+    struct Wildcard
+    {
+    };
+
+    ServiceDescriptionContainer_t m_serviceDescriptions;
+
+    // store the last known free Index (if any is known)
+    // we should not use a queue (or stack) here since they are not optimal
+    // for the filling pattern of a vector (prefer entries close to the front)
+    uint32_t m_freeIndex{NO_INDEX};
+
+  private:
+    // functions for the different search cases
+    // tag types are not needed in all cases to distinguish overloads but kept for consistency
+
+    uint32_t find(const capro::ServiceDescription& serviceDescription) const noexcept;
```

Comment:
There are 10 overloads of `find` and it seems you can create three groups of them when you consider the behavior. This is hard. Could we please change this.

The user api `find` can be named as it is right now. 

This `find` is actually something like `getIndexOfServiceDescription`.

And the remaining finds, why not make one method out of it and name it `fillSearchResultVector`.
```cpp
void fillSearchResultVector(const cxx::optional<capro::IdString_t> & service, const cxx::optional<capro::IdString_t> & instance, const cxx::optional<capro::IdString_t> & event, ServiceDescriptionVector_t& searchResult);
```

In the end I think the code does not become much more complex since one could write it like:
```cpp
for(auto & entry : m_serviceDescriptions) {
  bool isEqual = static_cast<bool>(entry);
  isEqual &= (service) ? (entry->serviceDescription.getServiceIDString() == *service) : true;
  isEqual &= (instance) ? (entry->serviceDescription.getInstanceIDString() == *instance) : true;
  isEqual &= (event) ? (entry->serviceDescription.getEventIDString() == *event) : true;
  if ( isEqual ) {
    searchResult.emplace_back(*entry);
  }
}
```

This would be much less code and complexity as with this approach.

Reply:
For the the first part which returns an index I agree but for the second part I do not.

I will not check `isEqual &= (service)` every time I look at an entry and it is in no way clear the compiler can and will optimize it (I tend to think it will not). That is why I check it **once** and then invoke the specific case handler (i.e. overload of find). This is exactly what overloading is for. Furthermore it is event conceivable to have these overloads public for cases where we already know that a specific wildcard is needed (at compile time). This would require testing them in a public interface which I do not want at this point though.

Furthermore, I saw in more sophisticated approaches (with a search tree) that it might be required to have very different case handling. Having different overloads for different cases makes this way easier.

TLDR: The code you present may look more compact but at the cost of being slower by doing unnecessary work and making it harder to change individual cases. Rejected for this reason.

Reply:
I disagree, lets ask @elBoberido as referee.

Reply:
@MatthiasKillat @elfenpiff As far as I know, this is more or less a stop gap solution right? When we notice that the search is too slow we would use the `PrefixTree` anyway? Is this assumption correct?
Furthermore, even if there is not yet a consensus on #1052, my bet would be that we will go with that route.

With all of this in mind, this looks like premature optimization and might even be removed in the not too distant future. While I understand the pursuit for performance, there are far more (yes, plural) elephants in the room which could give us a far bigger performance boost than this micro optimizations. Cases like false sharing, data distributed over two cache lines, context switches, etc. So all the stuff where the CPU needs to wait because of data synchronization or data not available. Stuff like that is more than an order of magnitude more relevant for performance issues than the stuff we constantly argue about. Please take a glass of wine and carefully read this blog post https://pkolaczk.github.io/server-slower-than-a-laptop/. This is the stuff we need to worry about.

So I'm leaning towards @elfenpiff since it tries to optimize something which does not even cause problems but this optimization adds costs in the form of maintenance and might make refactoring even harder.

Reply:
@elBoberido I am not sure it really will be a stop gap solution. And yes, we should strive to use atomics correctly as well etc. The #1052 would make the tree more attractive as well. Or a variation of it. But the tree is necessarily more complicated.

Reply:
@elfenpiff Changed under the condition that we (strongly) consider using the prefix tree for this and similar problems in the future. I dislike to use suboptimal (runtime etc.) algorithms just because they are simple. This is often deceiving and minor inefficiencies add up.